src/sentry/static/sentry/app/components/gridEditable/grid.tsx

```diff
@@ -0,0 +1,282 @@
+   * - `columnSortBy` is not used at the moment, however it might be better to
+   *   move sorting into Grid for performance
+   */
+  columnOrder: GridColumnOrder<ColumnKey>[];
+  columnSortBy: GridColumnSortBy<ColumnKey>[];
+  data: DataRow[];
+
+  /**
+   * GridEditable allows the parent component to determine how to display the
+   * data within it. Note that this is optional.
+   */
+  grid: {
+    renderHeaderCell?: {(column: GridColumnOrder<ColumnKey>): React.ReactNode};
+    renderBodyCell?: {
+      (column: GridColumnOrder<ColumnKey>, dataRow: DataRow): React.ReactNode;
+    };
+  };
+
+  /**
+   * As GridEditable is unopinionated about the structure of GridColumn,
+   * ModalEditColumn relys on the parent component to provide the form layout
+   * and logic to mutate the columns
+   */
+  modalEditColumn: {
+    renderBodyWithForm: {(column?: GridColumn<ColumnKey>): React.ReactNode};
+    renderFooter: {(): React.ReactNode};
+  };
+
+  /**
+   * As there is no internal state being maintained, the parent component will
+   * have to provide functions to mutate the column props
+   */
+  actions: {
+    moveColumn: {(fromIndex: number, toIndex: number): void};
+    deleteColumn: {(index: number): void};
+  };
+};
+export type GridEditableState = {
+  isEditing: boolean;
+  numColumn: number;
+};
+
+class GridEditable<DataRow extends Object> extends React.Component<
+  GridEditableProps<DataRow, keyof DataRow>,
+  GridEditableState
+> {
+  static defaultProps = {
+    isEditable: false,
+  };
+
+  state = {
+    numColumn: 0,
+    isEditing: false,
+  };
+
+  // Static methods do not allow the use of generics bounded to the parent class
+  // For more info: https://github.com/microsoft/TypeScript/issues/14600
+  static getDerivedStateFromProps(
+    props: GridEditableProps<Object, keyof Object>,
+    prevState: GridEditableState
+  ): GridEditableState {
+    return {
+      ...prevState,
+      numColumn: props.columnOrder.length,
+    };
+  }
+
+  toggleEdit = () => {
+    this.setState({isEditing: !this.state.isEditing});
+  };
+
+  toggleModalEditColumn = () => {
+    if (this.state.isEditing) {
+      const {modalEditColumn} = this.props;
+
+      openModal(openModalProps => (
+        <GridModalEditColumn
+          {...openModalProps}
+          renderBody={modalEditColumn.renderBodyWithForm}
+          renderFooter={modalEditColumn.renderFooter}
+        />
+      ));
+    }
+  };
+
+  renderError = () => {
+    const {error: errorMessage} = this.props;
+
+    return (
+      <React.Fragment>
+        <Alert type="error" icon="icon-circle-exclamation">
+          {errorMessage}
+        </Alert>
+        <Panel>
+          <Grid numColumn={this.state.numColumn}>{this.renderGridHead()}</Grid>
+        </Panel>
+      </React.Fragment>
+    );
+  };
+
+  renderLoading = () => {
+    return (
+      <GridBody>
+        <div style={{minHeight: '240px', gridColumn: '1 / -1'}}>
```

Comment:
Should we use a styled component for this and the empty data state and make them the same size?